File: examples/polaris/typescript/ci-fix-bot.ts
import { environment, outputText, stableKey, streamUntilTerminal } from "./common";

const jobId = process.env.CI_JOB_ID?.trim() || process.env.GITHUB_RUN_ID?.trim() || "local-ci-run";
const repository =
  process.env.CI_REPOSITORY?.trim() || process.env.GITHUB_REPOSITORY?.trim() || "repository";
const failureSummary =
  process.env.CI_FAILURE_SUMMARY?.trim() ||
  "The test job failed. Reproduce the failure, make the smallest safe repair, and verify it.";

const { polaris, projectId } = environment();
const session = await polaris.sessions.create(
  {
    projectId,
    title: `Repair CI for ${repository} (${jobId})`,
    provider: process.env.POLARIS_PROVIDER?.trim() || "codex",
    model: process.env.POLARIS_MODEL?.trim() || undefined,
  },
  { idempotencyKey: stableKey("ci-fix-session", repository, jobId) },
);

await session.sendTurn(
  {
    inputText: [
      `Repair the failed CI job ${jobId} in ${repository}.`,
      failureSummary,
      "Inspect the current repository state, preserve unrelated changes, run the focused failing test, then run the relevant quality gates.",
    ].join("\n\n"),
    runtimeMode: "full-access",
    interactionMode: "default",
  },
  { idempotencyKey: stableKey("ci-fix-turn", repository, jobId) },
);

const terminal = await streamUntilTerminal(session, {
  onEvent(event) {
    const text = outputText(event);
    if (text) process.stdout.write(text);
    if (
      event.eventType === "approval.requested" ||
      (event.eventType === "request.opened" &&
        typeof event.payload.requestType === "string" &&
        event.payload.requestType.endsWith("_approval"))
    ) {
      console.error("\nAgent is waiting for an approval; review it in the Polaris Console.");
    }
  },
});

if (terminal.eventType !== "execution.completed") {
  throw new Error(`CI repair ended with ${terminal.eventType}.`);
}
console.log(`\nCI repair Session completed: ${session.id}`);

File: examples/polaris/python/ci_fix_bot.py
from __future__ import annotations

import hashlib
import os

from polaris_agents import Polaris


def stable_key(kind: str, external_id: str) -> str:
    digest = hashlib.sha256(f"{kind}:{external_id}".encode()).hexdigest()
    return f"ci-{digest[:48]}"


polaris = Polaris(
    api_key=os.environ["POLARIS_API_KEY"],
    base_url=os.environ["POLARIS_BASE_URL"],
)
job_id = os.environ["CI_JOB_ID"]
session = polaris.sessions.create(
    project_id=os.environ["POLARIS_PROJECT_ID"],
    title=f"Fix CI job {job_id}",
    provider="codex",
    idempotency_key=stable_key("session", job_id),
)
session.send_turn(
    input_text="Reproduce the failing CI job, implement the smallest safe fix, and run focused tests.",
    idempotency_key=stable_key("turn", job_id),
)

for event in session.events():
    print(event["sequence"], event["eventType"])
    if event["eventType"] in {"turn.completed", "execution.failed"}:
        break
